src/hyperlight_host/src/hypervisor/mod.rs

```diff
@@ -523,322 +575,84 @@ pub trait InterruptHandle: Debug + Send + Sync {
-    /// **Bits 0-62**: Generation counter (incremented once per guest function call)
+    /// Bit layout:
+    /// - Bit 1: RUNNING_BIT - set when vcpu is actively running
+    /// - Bit 0: CANCEL_BIT - set when cancellation has been requested
     ///
-    /// # Generation Tracking
-    ///
-    /// The generation counter is incremented once at the start of each guest function call
-    /// and remains constant throughout that call, even if the VCPU is run multiple times
-    /// (due to host function calls, retries, etc.). This design solves the race condition
-    /// where a kill() from a previous call could spuriously cancel a new call.
-    ///
-    /// ## Why Generations Are Needed
-    ///
-    /// Consider this scenario WITHOUT generation tracking:
-    /// 1. Thread A starts guest call 1, VCPU runs
-    /// 2. Thread B calls kill(), sends signal to Thread A
-    /// 3. Guest call 1 completes before signal arrives
-    /// 4. Thread A starts guest call 2, VCPU runs again
-    /// 5. Stale signal from step 2 arrives and incorrectly cancels call 2
-    ///
-    /// WITH generation tracking:
-    /// 1. Thread A starts guest call 1 (generation N), VCPU runs
-    /// 2. Thread B calls kill(), stamps cancel_requested with generation N
-    /// 3. Guest call 1 completes, signal may or may not have arrived yet
-    /// 4. Thread A starts guest call 2 (generation N+1), VCPU runs again
-    /// 5. If stale signal arrives, signal handler checks: cancel_requested.generation (N) != current generation (N+1)
-    /// 6. Stale signal is ignored, call 2 continues normally
-    ///
-    /// ## Per-Call vs Per-Run Generation
-    ///
-    /// It's critical that generation is incremented per GUEST FUNCTION CALL, not per vcpu.run():
-    /// - A single guest function call may invoke vcpu.run() multiple times (host calls, retries)
-    /// - All run() calls within the same guest call must share the same generation
-    /// - This ensures kill() affects the entire guest function call atomically
-    ///
-    /// # Invariants
-    ///
-    /// - If VCPU is running: bit 63 is set (neither converse nor inverse holds)
-    /// - If VCPU is running: bits 0-62 match the current guest call's generation
-    running: AtomicU64,
+    /// CANCEL_BIT persists across vcpu exits/re-entries within a single `HyperlightVm::run()` call
+    /// (e.g., during host function calls), but is cleared at the start of each new `HyperlightVm::run()` call.
+    state: AtomicU64,
 
-    /// Thread ID where the VCPU is currently running.
-    ///
-    /// # Invariants
+    /// Thread ID where the vcpu is running.
     ///
-    /// - If VCPU is running: tid contains the thread ID of the executing thread
-    /// - Multiple VMs may share the same tid, but at most one will have running=true
+    /// Note: Multiple VMs may have the same `tid` (same thread runs multiple sandboxes sequentially),
+    /// but at most one VM will have RUNNING_BIT set at any given time.
```

Comment:
How do we know that the cancel was sent to the right vm in this case?  How will the work if the we are running multiple guest running in async manner?

Reply:
I don't think we can make that difference in this implementation, right?
I think that if we really want to make this difference, we can add some kind of identification for the running sandbox.

Reply:
Ok, for now I think this works but we may need to reconsider in scenarios where multiple guests might be doing async host calls